### PR TITLE
[SPARK-48505][CORE] Simplify the implementation of `Utils#isG1GC`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This PR changes to use the result of `ManagementFactory.getGarbageCollectorMXBeans` to determine whether G1GC is used. When G1GC is used, `ManagementFactory.getGarbageCollectorMXBeans` will return two instances of `GarbageCollectorExtImpl`, their names are `G1 Young Generation` and `G1 Old Generation` respectively.

### Why are the changes needed?
Simplify the implementation.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -3057,16 +3057,8 @@ private[spark] object Utils
...
-      val vmOptionClazz = Utils.classForName("com.sun.management.VMOption")
-      val hotSpotDiagnosticMXBean = ManagementFactory.getPlatformMXBean(clazz)
-      val vmOptionMethod = clazz.getMethod("getVMOption", classOf[String])
-      val valueMethod = vmOptionClazz.getMethod("getValue")
-
-      val useG1GCObject = vmOptionMethod.invoke(hotSpotDiagnosticMXBean, "UseG1GC")
-      val useG1GC = valueMethod.invoke(useG1GCObject).asInstanceOf[String]
-      "true".equals(useG1GC)
+      ManagementFactory.getGarbageCollectorMXBeans.asScala
+        .exists(_.getName.contains("G1"))
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3061]: friendly ping @dongjoon-hyun 

Do you think this simplification is ok? When using G1 GC, `ManagementFactory.getGarbageCollectorMXBeans` will return two instances of `GarbageCollectorExtImpl`, their names are `G1 Young Generation` and `G1 Old Generation` respectively. Thanks ~

**Comment by LuciferYang** [comment]
also cc @yaooqinn FYI

**Review by beliefer — COMMENTED** [review]
Looks a good improvement. We can avoid the Java reflection.

**Comment by yaooqinn** [comment]
It's not my area. So basically, I rely on the question below to be properly answered.

> How was this patch tested?

Can we point out the test case that covers it?

**Comment by LuciferYang** [comment]
@yaooqinn Although the logic is a bit convoluted, I am confident that the following two cases cover the scenarios where isG1GC is true and false, respectively.

https://github.com/apache/spark/blob/8d9d9c243ed141660e853b189982979d19f0ba9d/resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/BasicTestsSuite.scala#L145-L159



These two cases executed `org.apache.spark.examples.DriverSubmissionTest`, intercepted the logs, and checked the return value of `Utils.isG1GC` in the logs.

https://github.com/apache/spark/blob/8d9d9c243ed141660e853b189982979d19f0ba9d/resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala#L306-L343

https://github.com/apache/spark/blob/8d9d9c243ed141660e853b189982979d19f0ba9d/resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala#L656

https://github.com/apache/spark/blob/8d9d9c243ed141660e853b189982979d19f0ba9d/examples/src/main/scala/org/apache/spark/examples/DriverSubmissionTest.scala#L46


and they passed

image

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Thank you @LuciferYang , merged to master

**Comment by LuciferYang** [comment]
Thanks @yaooqinn

**Comment by rednaxelafx** [comment]
Thanks for the ping early on. I was holding back on making a comment to avoid derailing the thread, but thought I'd share some thoughts still.

I'm neutral to this change. It's a change that goes from one implementation-dependent code to another one of a different shape, the net-net is that both works in practice but it's hard to argue which one is superior. For the sake of having less code and avoiding reflection, I'm okay with the newly merged code, but I'd like to make sure we understand the caveats.
- the old code
  - Cons: uses reflection to probe the existence of the HotSpot VM specific `HotSpotDiagnosticMXBean`; but it's not _too_ bad, at least it doesn't require `setAccessible(true)`...
  - Pros: the -XX:+UseG1GC flag is precise and there's no room for any ambiguity
- the new code
  - Pros: plain Java code, no reflection
  - Cons: there's no hard contract/guarantee that this will match and will only match G1 GC; it's not _too_ bad in practice, that at least for now it'll work just fine on current commonly used JVMs

The string matching reminds me of some test code that used to be in the JDK that matched the exact exception string from certain exceptions (which wasn't a hard contract that should have been depended on), and when the JVM side made some optimizations that changed the string's spelling, that test was broken. You can argue either way.

The old code's initial attempt to use the `HotSpotDiagnosticMXBean` didn't use reflection, but rather it used the APIs directly, and later changed from direct use to reflection-based used to allow it to run on all JVMs. In fact, the very first commit in the original PR had pretty much the exact same code as the new one proposed here 😓 
https://github.com/apache/spark/pull/34846/commits/3a4c66117b157a8a73cc32deab0dc4e8c632db38#diff-2ecc6aef4b0c50bbf146e6c0b3b8b2249375f06a83e2a224c7718cfc850c3af7R3238

It's already very platform dependent when you're trying to do G1GC detection; the use of `HotSpotDiagnosticMXBean` falls in the category of "Using Oracle JDK's Platform Extension" ([JDK doc](https://docs.oracle.com/javase/8/docs/technotes/guides/management/mxbeans.html#gdfbk), JDK 17's doc also still links back to this JDK 8 page). I'd say I'm okay with the old reflection-based version for its preciseness.
It was using reflection all the way, but it's also possible to only do an initial reflection probe and then conditionally do direct use of `HotSpotDiagnosticMXBean` so that it "looks" less bloated.

**Comment by LuciferYang** [comment]
> Thanks for the ping early on. I was holding back on making a comment to avoid derailing the thread, but thought I'd share some thoughts still.
> 
> I'm neutral to this change. It's a change that goes from one implementation-dependent code to another one of a different shape, the net-net is that both works in practice but it's hard to argue which one is superior. For the sake of having less code and avoiding reflection, I'm okay with the newly merged code, but I'd like to make sure we understand the caveats.
> 
> * the old code
>   
>   * Cons: uses reflection to probe the existence of the HotSpot VM specific `HotSpotDiagnosticMXBean`; but it's not _too_ bad, at least it doesn't require `setAccessible(true)`...
>   * Pros: the -XX:+UseG1GC flag is precise and there's no room for any ambiguity
> * the new code
>   
>   * Pros: plain Java code, no reflection
>   * Cons: there's no hard contract/guarantee that this will match and will only match G1 GC; it's not _too_ bad in practice, that at least for now it'll work just fine on current commonly used JVMs
> 
> The string matching reminds me of some test code that used to be in the JDK that matched the exact exception string from certain exceptions (which wasn't a hard contract that should have been depended on), and when the JVM side made some optimizations that changed the string's spelling, that test was broken. You can argue either way.
> 
> The old code's initial attempt to use the `HotSpotDiagnosticMXBean` didn't use reflection, but rather it used the APIs directly, and later changed from direct use to reflection-based used to allow it to run on all JVMs. In fact, the very first commit in the original PR had pretty much the exact same code as the new one proposed here 😓 [3a4c661#diff-2ecc6aef4b0c50bbf146e6c0b3b8b2249375f06a83e2a224c7718cfc850c3af7R3238](https://github.com/apache/spark/commit/3a4c66117b157a8a73cc32deab0dc4e8c632db38#diff-2ecc6aef4b0c50bbf146e6c0b3b8b2249375f06a83e2a224c7718cfc850c3af7R3238)
> 
> It's already very platform dependent when you're trying to do G1GC detection; the use of `HotSpotDiagnosticMXBean` falls in the category of "Using Oracle JDK's Platform Extension" ([JDK doc](https://docs.oracle.com/javase/8/docs/technotes/guides/management/mxbeans.html#gdfbk), JDK 17's doc also still links back to this JDK 8 page). I'd say I'm okay with the old reflection-based version for its preciseness. It was using reflection all the way, but it's also possible to only do an initial reflection probe and then conditionally do direct use of `HotSpotDiagnosticMXBean` so that it "looks" less bloated.

@rednaxelafx Thank you very much for your explanation, based on your explanation, I think we should roll back this pr. 

WDYT @yaooqinn @beliefer ?

**Comment by LuciferYang** [comment]
> It was using reflection all the way, but it's also possible to only do an initial reflection probe and then conditionally do direct use of HotSpotDiagnosticMXBean so that it "looks" less bloated.

@rednaxelafx I've submitted a PR to try and make further refactor according to this suggestion. If you have time, could you please take a look?

- https://github.com/apache/spark/pull/46873/files

**Comment by yaooqinn** [comment]
Thank you @rednaxelafx for the detailed inputs.

I reverted this with https://github.com/apache/spark/commit/db527ac346f2f6f6dbddefe292a24848d1120172, since it's controversial 

Let's discuss further in https://github.com/apache/spark/pull/46873 @LuciferYang

**Comment by LuciferYang** [comment]
Thanks @yaooqinn , I reopened SPARK-48505

**Pull request closed without merging**